File: top_k_frequent/top_k_frequent.ts
export {};

function topKFrequent(nums: number[], k: number): number[] {
  const frequencyMap = new Map<number, number>();
  for (const num of nums) {
    frequencyMap.set(num, (frequencyMap.get(num) || 0) + 1);
  }
  const sortedEntries = [...frequencyMap.entries()].sort((a, b) => b[1] - a[1]);
  console.log('Frequency Map:', frequencyMap);
  console.log('Sorted Entries:', sortedEntries);
  return sortedEntries.slice(0, k).map(entry => entry[0]);
}

const nums = [1, 1, 1, 2, 2, 3, 5, 5, 5, 5, 6, 6, 7, 8, 8, 8, 8];
const k = 2;

const result = topKFrequent(nums, k);
console.log(`Top ${k} frequent elements: `, result);

File: top_k_frequent/top_k_frequent.py
from collections import Counter

def topKFrequent(nums, k):
    frequency_map = {}
    for num in nums:
        frequency_map[num] = frequency_map.get(num, 0) + 1
    sorted_entries = sorted(frequency_map.items(), key=lambda x: x[1], reverse=True)
    return [entry[0] for entry in sorted_entries[:k]]   

def topKFrequentUsingCounter(nums, k):
    frequency_map = Counter(nums)
    common = frequency_map.most_common(k)
    return [element for element, _ in common]

nums = [1, 1, 1, 2, 2, 3]   
k = 2
result = topKFrequent(nums, k)
counter_result = topKFrequentUsingCounter(nums, k)

print("The top", k, "frequent elements are:", result);
print("Using Counter:", counter_result);